tests/test_dataset_common.py

```diff
@@ -322,7 +322,9 @@ def test_builder_configs(self, dataset_name):
 
 def distributed_load_dataset(args):
     data_name, tmp_dir, datafiles = args
-    dataset = load_dataset(data_name, cache_dir=tmp_dir, data_files=datafiles)
+    # Disable multiprocessing in download to avoid AssertionError: daemonic processes are not allowed to have children
+    download_config = DownloadConfig(num_proc=1)
+    dataset = load_dataset(data_name, cache_dir=tmp_dir, data_files=datafiles, download_config=download_config)
```

Comment:
I think this could break some example scripts in `transformers` that use pytorch DDP. Switching to multithreading could help maybe, which is more appropriate to download files anyway.

Reply:
@lhoestq the intention of this PR was not to change the downloading mechanism, but to address the request by @nateraw:
```
nateraw
Maybe it would be nice if num_proc > num_files then you set num_proc = num_files.

Quentin Lhoest
Yep indeed
```

The specific use case was that when passing `num_proc=20` but having 19 files to download, multiprocessing was not used (which is counter-intuitive).

From the user point of view, I have raised my concerns in this PR description: https://github.com/huggingface/datasets/pull/4740#issue-1316478007

Reply:
I see, I'm just a bit concerned with the regression it might cause for users that use multiprocessing already e.g. to train models.

Reply:
@lhoestq I agree this is a breaking change, thus my doubts as well, as explained at the end of the PR description: https://github.com/huggingface/datasets/pull/4740#issue-1316478007

My usual preference would have been being explicit (as the Zen of Python `Explicit is better than implicit.`), thus I would have recommended not doing multiprocessing unless explicitly required by the user.

However, given that this implicit multiprocessing is already in place (when more than 16 files to download), IMO is preferable to keep backward compatibility (thus not introducing breaking changes) as much as possible.

I'm thinking of a workaround...

Reply:
> My usual preference would have been being explicit (as the Zen of Python Explicit is better than implicit.), thus I would have recommended not doing multiprocessing unless explicitly required by the user.

100% agree, though using multithreading to download files in parallel is something that doesn't need user's permission IMO. Multiprocessing has some side effects so I agree it's better if users ask for it explicitly. That's why I think we should switch to multithreading at one point